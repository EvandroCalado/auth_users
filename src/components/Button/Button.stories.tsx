import Button, { ButtonProps } from './Button';
import { Meta, StoryFn } from '@storybook/react';
import { VpnKey } from '@styled-icons/material-outlined';

export default {
  title: 'Button',
  component: Button,
  args: {
    children: 'Bot√£o',
  },
} as Meta;

export const WithoutIcon: StoryFn<ButtonProps> = (args) => (
  <div>
    <Button {...args} />
  </div>
);

export const WithIcon: StoryFn<ButtonProps> = (args) => (
  <div>
    <Button {...args} icon={<VpnKey />} />
  </div>
);

export const SecondaryColor: StoryFn<ButtonProps> = (args) => (
  <div>
    <Button {...args} color="secondary" icon={<VpnKey />} />
  </div>
);

export const Disabled: StoryFn<ButtonProps> = (args) => (
  <div>
    <Button {...args} icon={<VpnKey />} disabled={true} />
  </div>
);
