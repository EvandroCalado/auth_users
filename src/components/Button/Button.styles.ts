import styled, { css } from 'styled-components';
import { ButtonProps } from './Button';

export const Button = styled.button<Pick<ButtonProps, 'color'>>`
  ${({ theme, color }) => css`
    background-color: ${theme.colors[color]};
    color: ${theme.colors.white};
    font-size: ${theme.font.sizes.small};
    border: none;
    padding: ${theme.spacings.xsmall} ${theme.spacings.medium};
    border-radius: ${theme.spacings.tiny};
    transition: ${theme.transitions.fast};
    display: flex;
    align-items: center;
    justify-content: center;
    cursor: pointer;

    &:focus {
      outline: none;
      box-shadow: 0 0 ${theme.spacings.tiny} ${theme.colors[color]};
      filter: brightness(110%);
    }

    &:hover {
      filter: brightness(90%);
    }

    &:disabled {
      background-color: ${theme.colors.gray4};
      color: ${theme.colors.gray1};
      cursor: not-allowed;

      &:hover {
        filter: none;
      }

    }
    
    > svg {
      width: 2rem;
      height: 2rem;
      margin-left: 1rem;
    }
  `}
`;
