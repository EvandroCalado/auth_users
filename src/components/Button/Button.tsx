import { ButtonHTMLAttributes } from 'react';
import * as Styled from './Button.styles';

export type ButtonProps = {
  children: React.ReactNode;
  disabled?: boolean;
  onClick?: () => void;
  icon?: React.ReactNode;
  color?: 'primary' | 'secondary';
} & ButtonHTMLAttributes<HTMLButtonElement>;

const Button = ({
  children,
  disabled = false,
  onClick,
  icon,
  color = 'primary',
}: ButtonProps) => {
  const handleClick = () => {
    if (onClick) {
      onClick();
    }
  };

  return (
    <Styled.Button disabled={disabled} onClick={handleClick} color={color}>
      {children}
      {icon && icon}
    </Styled.Button>
  );
};

export default Button;
