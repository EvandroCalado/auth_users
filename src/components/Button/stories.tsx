import Button, { ButtonProps } from '.';
import { Meta, StoryFn } from '@storybook/react';
import { VpnKey } from '@styled-icons/material-outlined';

export default {
  title: 'Button',
  component: Button,
  args: {
    children: 'Texto do Bot√£o',
    icon: <VpnKey />,
  },
} as Meta;

export const Template: StoryFn<ButtonProps> = (args) => (
  <div>
    <Button {...args} />
  </div>
);
