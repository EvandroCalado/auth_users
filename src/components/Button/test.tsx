import Button from '.';
import { fireEvent } from '@testing-library/react';
import { renderTheme } from '../../styles/render-theme';
import { VpnKey } from '@styled-icons/material-outlined';
import { vi } from 'vitest';

test('renders a disabled button when disabled prop is true', () => {
  const handleClick = vi.fn();
  const { getByRole } = renderTheme(
    <Button
      disabled={true}
      onClick={handleClick}
      color="primary"
      icon={<VpnKey />}
    >
      Click me
    </Button>,
  );
  const button = getByRole('button');

  expect(button).toBeDisabled();
});

test('calls onClick handler when button is clicked', () => {
  const handleClick = vi.fn();
  const { getByRole } = renderTheme(
    <Button onClick={handleClick} color="primary" icon={<VpnKey />}>
      Click me
    </Button>,
  );
  const button = getByRole('button');

  fireEvent.click(button);

  expect(handleClick).toHaveBeenCalled();
});

test('renders with the correct color', () => {
  const { getByRole } = renderTheme(
    <Button color="primary" icon={<VpnKey />}>
      Click me
    </Button>,
  );
  const button = getByRole('button');

  expect(button).toContainHTML('primary');
});

test('renders with the icon', () => {
  const { getByRole } = renderTheme(
    <Button color="primary" icon={<VpnKey />}>
      Click me
    </Button>,
  );
  const button = getByRole('button');
  const icon = button.lastChild;

  expect(icon).toBeInTheDocument();
});

test('to match snapshot', () => {
  const { container } = renderTheme(
    <Button color="primary" icon={<VpnKey />}>
      Click me
    </Button>,
  );
  expect(container).toMatchSnapshot();
});
