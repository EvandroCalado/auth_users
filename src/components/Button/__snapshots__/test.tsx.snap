// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`to match snapshot 1`] = `
.c0 {
  background-color: rgb(43,38,178);
  border: none;
  color: rgb(250,250,253);
  padding: 1.2rem 2.4rem;
  cursor: pointer;
  border-radius: 0.4rem;
  -webkit-transition: all 300ms ease-in-out;
  transition: all 300ms ease-in-out;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0:focus {
  outline: none;
  box-shadow: 0 0 0.4rem rgb(43,38,178);
  -webkit-filter: brightness(110%);
  filter: brightness(110%);
}

.c0:hover {
  -webkit-filter: brightness(90%);
  filter: brightness(90%);
}

.c0:disabled {
  background-color: rgb(181,181,184);
  color: rgb(241,241,243);
  cursor: not-allowed;
}

.c0:disabled:hover {
  -webkit-filter: none;
  filter: none;
}

.c0 > svg {
  width: 2rem;
  height: 2rem;
  margin-left: 1rem;
}

.c1 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

<div>
  <button
    class="c0"
    color="primary"
  >
    Click me
    <svg
      aria-hidden="true"
      class="c1"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M22 19h-6v-4h-2.68c-1.14 2.42-3.6 4-6.32 4-3.86 0-7-3.14-7-7s3.14-7 7-7c2.72 0 5.17 1.58 6.32 4H24v6h-2v4zm-4-2h2v-4h2v-2H11.94l-.23-.67C11.01 8.34 9.11 7 7 7c-2.76 0-5 2.24-5 5s2.24 5 5 5c2.11 0 4.01-1.34 4.71-3.33l.23-.67H18v4zM7 15c-1.65 0-3-1.35-3-3s1.35-3 3-3 3 1.35 3 3-1.35 3-3 3zm0-4c-.55 0-1 .45-1 1s.45 1 1 1 1-.45 1-1-.45-1-1-1z"
      />
    </svg>
  </button>
</div>
`;
